sdks/python/apache_beam/io/aws/clients/s3/boto3_client.py

```diff
@@ -67,6 +69,11 @@ def __init__(self, options):
         aws_secret_access_key=secret_access_key,
         aws_session_token=session_token)
 
+    self._download_request = None
```

Comment:
I haven't had any issues with that on a Flink backend. I don't know the SDK in detail, but I don't think it creates any issues, because the entire behaviour of the client is still idempotent (just less efficient if you don't read sequentially).

Reply:
if an instance would be shared to process multiple bundles in parallel, the result really becomes unpredictable. but it sounds like that's not the case

Reply:
i'm certainly not opposed to a stateful client here, i just lack the context if this can be safely done

Reply:
If an instance is used to process two connections simultaneously, the client should reset the connection before each range request and effectively fall back to the old behaviour. The same happens when you read non-sequentially. The only issue I can see here if the client is used in a non-thread-safe way, but I don't want to add mutex locks unless I know that this is definitely a scenario that can occur.

Reply:
well, these blocks are by no means atomic... there's all kinds of possible race conditions if this code should run concurrently

Reply:
👍